File: node.js/examples/OfflineProcessingTrie.js
/*
This Source Code Form is copyright of 51Degrees Mobile Experts Limited.
Copyright (c) 2017 51Degrees Mobile Experts Limited, 5 Charlotte Close,
Caversham, Reading, Berkshire, United Kingdom RG4 7BY

This Source Code Form is the subject of the following patent
applications, owned by 51Degrees Mobile Experts Limited of 5 Charlotte
Close, Caversham, Reading, Berkshire, United Kingdom RG4 7BY:
European Patent Application No. 13192291.6; and
United States Patent Application Nos. 14/085,223 and 14/085,301.

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0.

If a copy of the MPL was not distributed with this file, You can obtain
one at http://mozilla.org/MPL/2.0/.

This Source Code Form is "Incompatible With Secondary Licenses", as
defined by the Mozilla Public License, v. 2.0.
*/

/*
<tutorial>
Offline processing example of using 51Degrees device detection. The example
shows how to:
<ol>
<li>Set the configuration using a json object
<p><pre class="prettyprint lang-js">
var config = {"dataFile" : require("fiftyonedegreescore"),
              "properties" : "IsMobile",
              "cacheSize" : 10000,
              "poolSize" : 4
};
</pre></p>
<li>Instantiate the 51Degrees device detection provider with this
connfiguration
<p><pre class="prettyprint lang-js">
var provider = new fiftyonedegrees.provider(config);
</pre></p>
<li>Open an input file with a list of User-Agents, and an output file,
<p><pre class="prettyprint lang-js">
var inStream = fs.createReadStream(inputFile);
var outStream = fs.createWriteStream(outputFile);
</pre></p>
<li>Write a header to the output file with the property names in '|'
separated CSV format ('|' separated because some User-Agents contain
commas)
<p><pre class="prettyprint lang-js">
outStream.write('User-Agent')
provider.availableProperties.forEach(function(property) {
    outStream.write('|' + property);
});
outStream.write('\n');
</pre></p>
<li>For the first 20 User-Agents in the input file, perform a match then
write the User-Agent along with the values for chosen properties to
the CSV.
<p><pre class="prettyprint lang-js">
rl.on('line', function (userAgent) {
    if (i < 20) {
        outStream.write(userAgent);
        var match = provider.getMatch(userAgent);
        provider.availableProperties.forEach(function(property) {
            outStream.write("|" + match[property]);
        })
        outStream.write("\n");
        match.close();
    }
    i++;
});
</pre></p>
</ol>
</tutorial>
*/

// Snippet Start
// Include 51Degrees.
var fiftyonedegrees = require("fiftyonedegreescore");
// Inlcude file modules.
var fs = require("fs"),
    readline = require("readline");

// Set the config.
var config = {"dataFile" : require("fiftyonedegreeslitetrie"),
              "properties" : "IsMobile,PlatformName,PlatformVersion",
              "stronglyTyped" : false
             };

var inputFile = __dirname + "/../../data/20000 User Agents.csv";
var outputFile = "offlineProcessingOutput.csv";

// Carries out match for first 20 User-Agents and prints results to
// output file.
var outputOfflineProcessing = function() {
    var inStream = fs.createReadStream(inputFile),
        rl = readline.createInterface(inStream, null),
        outStream = fs.createWriteStream(outputFile);
    inStream.on("close", function() {
        outStream.close();
        console.log("Output written to " + outputFile);
    })

    outStream.write("User-Agent");
    provider.availableProperties.forEach(function(property) {
        outStream.write("|" + property);
    })
    outStream.write("\n");

    var i = 0;
    rl.on('line', function (userAgent) {
        if (i < 20) {
            outStream.write(userAgent);
            var match = provider.getMatch(userAgent);
            provider.availableProperties.forEach(function(property) {
                outStream.write("|" + match[property]);
            })
            outStream.write("\n");
            match.close();
        }
        i++;
    });
}

console.log("Starting Offline Processing Example.\n");

/*
Initialises the device detection provider with settings from the config.
By default this will use the packaged Lite data file. For more info see:
<a href="https://51degrees.com/compare-data-options">compare data options
</a>
*/
var provider = new fiftyonedegrees.provider(config);
outputOfflineProcessing();
// Snippet End

File: python/examples/OfflineProcessingTrie.py
#!/usr/bin/env python
'''
This Source Code Form is copyright of 51Degrees Mobile Experts Limited.
Copyright 2017 51Degrees Mobile Experts Limited, 5 Charlotte Close,
Caversham, Reading, Berkshire, United Kingdom RG4 7BY

This Source Code Form is the subject of the following patents and patent
applications, owned by 51Degrees Mobile Experts Limited of 5 Charlotte
Close, Caversham, Reading, Berkshire, United Kingdom RG4 7BY:
European Patent No. 2871816;
European Patent Application No. 17184134.9;
United States Patent Nos. 9,332,086 and 9,350,823; and
United States Patent Application No. 15/686,066.

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0.

If a copy of the MPL was not distributed with this file, You can obtain
one at http://mozilla.org/MPL/2.0/.

This Source Code Form is "Incompatible With Secondary Licenses", as
defined by the Mozilla Public License, v. 2.0.
'''

'''
<tutorial>
Offline processing example of using 51Degrees device detection. The example
shows how to:
<ol>
<li>Import settings from the 51Degrees settings file
<p><pre class="prettyprint lang-py">
dataFile = settings.V3_TRIE_WRAPPER_DATABASE
properties = settings.PROPERTIES
</pre></p>
<li>Instantiate the 51Degrees device detection provider with these
properties
<p><pre class="prettyprint lang-py">
provider = fiftyone_degrees_mobile_detector_v3_trie_wrapper.Provider(dataFile,
	properties)
</pre></p>
<li>Open an input file with a list of User-Agents, and an output file,
<p><pre class="prettyprint lang-py">
fin = open(inputFile, 'r')
fout = open(outputFile, 'w')
</pre></p>
<li>Write a header to the output file with the property names in '|'
separated CSV format ('|' separated because some User-Agents contain
commas)
<p><pre class="prettyprint lang-py">
fout.write('User-Agent')
for name in properties.split(','):
	fout.write('|' + name)
fout.write('\n')
</pre></p>
<li>For the first 20 User-Agents in the input file, perform a match then
write the User-Agent along with the values for chosen properties to
the CSV.
<p><pre class="prettyprint lang-py">
while i < 20 :
	userAgent = fin.readline().rstrip('\n')
	fout.write(userAgent)
	device = provider.getMatch(userAgent)
	for name in properties.split(','):
		value = device.getValue(name)
		fout.write('|' + value)
	fout.write('\n')
	i = i + 1
</pre></p>
</ol>
This example can be run in any directory, but assumes your
settings file contains a valid dataFile location and has the
IsMobile property selected.
</tutorial>
'''
# // Snippet Start
from FiftyOneDegrees import fiftyone_degrees_mobile_detector_v3_trie_wrapper
from fiftyone_degrees.mobile_detector.conf import settings
import sys

'''
Imports settings from the settings file. The Default settings file, and
details on how to change it can be output by running the command
<p><pre class="prettyprint lang-py">
51degrees-mobile-detector settings
</p></pre>
'''
dataFile = settings.V3_TRIE_WRAPPER_DATABASE
properties = settings.PROPERTIES

inputFile = '../../data/20000 User Agents.csv'
outputFile = 'offlineProcessingOutput.csv'

'''
Initialises the device detection provider with settings from the settings
file. By default this will use the included Lite data file For more info
see:
<a href="https://51degrees.com/compare-data-options">compare data options
</a>
'''
provider = fiftyone_degrees_mobile_detector_v3_trie_wrapper.Provider(
	dataFile,
	properties)

# Carries out match for first 20 User-Agents and prints results to
# output file.
def output_offline_processing():
	fin = open(inputFile, 'r')
	fout = open(outputFile, 'w')
	fout.write('User-Agent')
	for name in properties.split(','):
		fout.write('|' + name)

	fout.write('\n')
	i = 0
	while i < 20 :
		userAgent = fin.readline().rstrip('\n')
		fout.write(userAgent)
		match = provider.getMatch(userAgent)
		for name in properties.split(','):
			value = match.getValue(name)
			fout.write('|' + value)
		fout.write('\n')
		i += 1
	fin.close()
	fout.close()

def main():
	sys.stdout.write('Starting Offline Processing\n')
	output_offline_processing()
	sys.stdout.write('Output Written to %s\n' % outputFile)
if __name__ == '__main__':
	main()
# // Snippet End
